packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -1296,22 +1303,73 @@ function dispatchAction<S, A>(
     update.priority = getCurrentPriorityLevel();
   }
 
-  // Append the update to the end of the list.
-  const pending = queue.pending;
-  if (pending === null) {
-    // This is the first update. Create a circular list.
-    update.next = update;
+  appendUpdate(queue, update);
+
+  const alternate = fiber.alternate;
+  if (
+    fiber === currentlyRenderingFiber ||
+    (alternate !== null && alternate === currentlyRenderingFiber)
+  ) {
+    // This is a render phase update. Stash it in a lazily-created map of
+    // queue -> linked list of updates. After this render pass, we'll restart
+    // and apply the stashed updates on top of the work-in-progress hook.
+    didScheduleRenderPhaseUpdate = true;
+    update.expirationTime = renderExpirationTime;
+    currentlyRenderingFiber.expirationTime = renderExpirationTime;
   } else {
-    update.next = pending.next;
-    pending.next = update;
+    if (__DEV__) {
+      // $FlowExpectedError - jest isn't a global, and isn't recognized outside of tests
+      if ('undefined' !== typeof jest) {
+        warnIfNotScopedWithMatchingAct(fiber);
+        warnIfNotCurrentlyActingUpdatesInDev(fiber);
+      }
+    }
+    scheduleWork(fiber, expirationTime);
+  }
+}
+
+function setState<S>(
```

Comment:
For now this is being split from the `dispatchAction` but they both share pretty much the same implementation - besides the fact that `dispatchAction` has no eager bailout implemented. 

Both could be collapsed into a single function and the eager bailout could only be attempted for `reducer === basicStateReducer` but to do that I would have to bring back keeping track of `lastRenderedReducer`. 

Let me know which approach do you prefer.